src/librbd/object_map/DiffRequest.cc

```diff
@@ -22,20 +23,38 @@ using util::create_rados_callback;
 
 template <typename I>
 void DiffRequest<I>::send() {
-  std::shared_lock image_locker{m_image_ctx->image_lock};
+  auto cct = m_image_ctx->cct;
 
-  m_diff_from_start = (m_snap_id_start == 0);
-  if (m_snap_id_start == 0) {
-    if (!m_image_ctx->snap_info.empty()) {
-      m_snap_id_start = m_image_ctx->snap_info.begin()->first;
-    } else {
-      m_snap_id_start = CEPH_NOSNAP;
-    }
+  std::optional<int> r = std::nullopt;
+  if (m_snap_id_start == CEPH_NOSNAP || m_snap_id_start > m_snap_id_end) {
+    lderr(cct) << "invalid start/end snap ids: "
+               << "snap_id_start=" << m_snap_id_start << ", "
+               << "snap_id_end=" << m_snap_id_end << dendl;
+    r = -EINVAL;
+  } else if (m_snap_id_start == m_snap_id_end) {
+    // no delta between the same snapshot
+    r = 0;
+  }
+
+  if (r) {
+    auto ctx = new LambdaContext([this](int r) { finish(r); });
+    m_image_ctx->op_work_queue->queue(ctx, *r);
```

Comment:
forgot about return?

Reply:
Indeed -- I guess I just need to finally add the unit test case for this state machine.